### PR TITLE
Add a transformer to drop outliers

### LINKED ISSUE #2704: Add `DropOutlierRowsTransformer`
https://github.com/alteryx/evalml/pull/2692 introduced a generic `DropRowsTransformer` but based on the thread [here](https://github.com/alteryx/evalml/issues/2100#issuecomment-905875874), it'd be a good idea to introduce a data-agnostic component which detects and drops outliers.

This could probably be done by subclassing `DropRowsTransformer` and adding outlier detection logic!

### PR DESCRIPTION
**Pull request by angela97lin**
Closes #2704

### CODE REVIEW

File: docs/source/api_index.rst [hunk 1]
```diff
@@ -410,13 +411,13 @@ Data Check Classes
...
     evalml.data_checks.IDColumnsDataCheck
     evalml.data_checks.TargetLeakageDataCheck
-    evalml.data_checks.OutliersDataCheck
     evalml.data_checks.NoVarianceDataCheck
     evalml.data_checks.ClassImbalanceDataCheck
     evalml.data_checks.MulticollinearityDataCheck
     evalml.data_checks.DateTimeFormatDataCheck
     evalml.data_checks.TimeSeriesParametersDataCheck
     evalml.data_checks.TimeSeriesSplittingDataCheck
+    evalml.data_checks.outliers_data_check.OutliersDataCheck
```
Comment (angela97lin) [inline docs/source/api_index.rst:420]: Another example of https://github.com/alteryx/evalml/issues/2548

Comment (angela97lin) [reply docs/source/api_index.rst:420]: (Failed build: https://readthedocs.com/projects/feature-labs-inc-evalml/builds/852750/)

File: evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+"""Transformer to drop rows specified by row indices."""
+from evalml.data_checks.outliers_data_check import OutliersDataCheck
+from evalml.pipelines.components.transformers import Transformer
+from evalml.utils import infer_feature_types
+
+
+class DropOutliersTransformer(Transformer):
```
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py:7]: I don't think it is a good idea to re-run the outlier calculation in this component. There is no guarantee that the same rows will be dropped when users go to re-train the dataset. This is the scenario I'm thinking about:

1. Users create a training/validation split
2. Users create pipelines with DropOutliersTransformers and pass those pipelines to AutoMLSearch
3. Outliers are calculated on each fold of the training set, those are dropped, and scores are calculated on the validation set.
4. Scores look good, user gets a pipeline with DropOutliersTransformer and calls fit on the entire dataset (training + validation concatted together).

In steps 3 and 4, different rows could identified as having outliers because the distribution of the data is slightly changing between the folds. I think it's problematic that when users fit on their entire data set, new rows could be dropped from the training set without their knowledge. 

So I think this component should not determine what an outlier is for users and then automatically drop. We should let users tell us what they think outliers are based on their knowledge of their own data. A way to do this is to accept an `outlier_ranges` parameter:

```python
outlier_ranges: {"col_name" {"low_bound": float, "high_bound": float}
```

Anything lower than `low_bound` or higher than `high_bound` is an outlier and that will be true for all datasets users train the pipeline on.

File: evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        all_rows_with_indices_set = set()
+        for row_indices in self.outlier_rows_dict.values():
+            all_rows_with_indices_set.update(row_indices)
+
+        all_rows_with_indices = list(all_rows_with_indices_set)
+        all_rows_with_indices.sort()
+        self.outlier_indices = all_rows_with_indices
+
+        X_t = X_t.drop(self.outlier_indices, axis=0)
+        X_t.ww.init()
```
Comment (freddyaboulton) [inline evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py:65]: This will reset the woodwork schema

Comment (jeremyliweishih) [inline evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py:65]: Would using `X_t.ww.drop()` instead make it so we don't need to re-init?

**Review by jeremyliweishih — APPROVED** [review]
looks great!

File: evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py (hunk 2 above)
Comment (angela97lin) [reply evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py:7]: @freddyaboulton Interesting thoughts on this! I see the concern you're raising. I still wonder if there's a more "automatic" approach to dropping outliers (or if that's even desirable). Having an upper and lower bound requires the user to pass in the information to AutoMLSearch but I wish there were a way to automatically detect what good values for that is.

Is the scenario we outlined not the same for other components such as an Imputer with an imputation strategy of "most_frequent" for example? Where when we train, we could impute with what is considered the most frequent value found during training but once we concatenate training + validation, we end up with a different most_frequent value and end up imputing using a different value?

Comment (freddyaboulton) [reply evalml/pipelines/components/transformers/preprocessing/drop_outliers_transformer.py:7]: Yea that's a great point. It is similar to the Imputer case but I think there's still a difference - the concept of an outlier is always relative to the other data you have collected. What you consider an outlier based on 1000 row sample could change if you collect another 1000 rows. The rows that are missing on the first 1000 row sample will still be missing if you add the next 1000 rows to the original sample. Of course, the value you use to fill in those missing values could change depending on the data you call `fit` on but the subjectiveness of outliers makes me wary of automatically dropping rows in pipeline fitting.

**Comment by angela97lin** [comment]
Closing with respect to https://github.com/alteryx/evalml/issues/2704#issuecomment-1040677347
